File: app/models/pedidos.py
from ..utils.db import db


class Pedidos(db.Model):
    __tablename__ = 'Pedidos'

    id = db.Column(db.Integer, primary_key=True)
    cliente_id = db.Column(db.Integer, db.ForeignKey('Clientes.id'), nullable=False)
    fecha_pedido = db.Column(db.DateTime, server_default=db.func.current_timestamp())
    fecha_entrega = db.Column(db.Date)
    estado = db.Column(db.Enum('pendiente', 'entregado', 'cancelado', name='estado_pedido'), default='pendiente')
    total = db.Column(db.Numeric(10, 2))

    detalles = db.relationship('DetallePedido', backref='pedido', lazy=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'cliente_id': self.cliente_id,
            'fecha_pedido': self.fecha_pedido.isoformat() if self.fecha_pedido else None,
            'fecha_entrega': self.fecha_entrega.isoformat() if self.fecha_entrega else None,
            'estado': self.estado,
            'total': float(self.total) if self.total else None,
            'detalles': [detalle.to_dict() for detalle in self.detalles] if self.detalles else []
        }


class DetallePedido(db.Model):
    __tablename__ = 'Detalles_pedido'

    id = db.Column(db.Integer, primary_key=True)
    pedido_id = db.Column(db.Integer, db.ForeignKey('Pedidos.id'), nullable=False)
    producto_id = db.Column(db.Integer, db.ForeignKey('Productos.id'), nullable=False)
    cantidad = db.Column(db.Integer, nullable=False, default=1)
    precio_unitario = db.Column(db.Numeric(10, 2), nullable=False)
    subtotal = db.Column(db.Numeric(10, 2), nullable=False)

    producto = db.relationship('Productos', backref='detalles', lazy=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'pedido_id': self.pedido_id,
            'producto_id': self.producto_id,
            'cantidad': self.cantidad,
            'precio_unitario': float(self.precio_unitario) if self.precio_unitario else None,
            'subtotal': float(self.subtotal) if self.subtotal else None,
            'producto': self.producto.to_dict() if self.producto else None
        } 

File: frontend/src/app/services/pedidos.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Producto } from './productos';

export interface PedidoDetalle {
  id: number;
  pedido_id: number;
  producto_id: number;
  cantidad: number;
  precio_unitario: number;
  subtotal: number;
  producto: Producto | null;
}

export interface Pedido {
  id: number;
  cliente_id: number;
  fecha_pedido: string | null;
  fecha_entrega: string | null;
  estado: 'pendiente' | 'entregado' | 'cancelado';
  total: number | null;
  detalles: PedidoDetalle[];
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  status_code: number;
  timestamp: string;
  errors: unknown;
  data: T;
}

export interface PedidosData {
  pedidos: Pedido[];
  pagina: number;
  por_pagina: number;
  total_paginas: number;
  total_pedidos: number;
}

export interface PedidoLineaPayload {
  producto_id: number;
  cantidad: number;
}

export interface CrearPedidoPayload {
  cliente_id: number;
  fecha_entrega?: string | null;
  productos_seleccionados: PedidoLineaPayload[];
}

export interface ActualizarPedidoPayload {
  fecha_entrega?: string | null;
  estado?: 'pendiente' | 'entregado' | 'cancelado';
  productos_seleccionados?: PedidoLineaPayload[];
}

@Injectable({
  providedIn: 'root'
})
export class PedidosService {
  private apiUrl = 'http://localhost:5000/pedidos/api/';

  constructor(private http: HttpClient) {}

  obtenerPedidos(
    pagina = 1,
    porPagina = 10,
    cliente = '',
    estado = ''
  ): Observable<ApiResponse<PedidosData>> {
    let params = new HttpParams()
      .set('pagina', pagina)
      .set('por_pagina', porPagina);

    if (cliente.trim()) {
      params = params.set('cliente', cliente.trim());
    }

    if (estado.trim()) {
      params = params.set('estado', estado.trim());
    }

    return this.http.get<ApiResponse<PedidosData>>(this.apiUrl, { params });
  }

  crearPedido(payload: CrearPedidoPayload): Observable<ApiResponse<Pedido>> {
    return this.http.post<ApiResponse<Pedido>>(this.apiUrl, payload);
  }

  actualizarPedido(id: number, payload: ActualizarPedidoPayload): Observable<ApiResponse<Pedido>> {
    return this.http.put<ApiResponse<Pedido>>(`${this.apiUrl}${id}`, payload);
  }
}
